src/main/scala/scalismo/faces/segmentation/LoopyBPSegmentation.scala

```diff
@@ -0,0 +1,478 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+package scalismo.faces.segmentation
+
+import java.io.File
+
+import scalismo.faces.color.{ColorDistribution, GaussDist, RGB, UniformDist}
+import scalismo.faces.image.{ImageBuffer, PixelImage}
+import scalismo.faces.gui.GUIBlock._
+import scalismo.faces.gui.{GUIFrame, ImagePanel}
+import scalismo.faces.io.PixelImageIO
+import scalismo.utils.Random
+
+import scala.collection.immutable.IndexedSeq
+
+
+// an implementation of loopy belief propagation for segmentation
+object LoopyBPSegmentation {
+
+
+  // the directions for message passing
+  private sealed trait Direction {
+    def toInt: Int
+  }
+
+  private object Direction {
+    val allDirections = IndexedSeq(Right, Left, Down, Up)
+
+    def fromIndex(index: Int): Direction = index match {
+      case 0 => Right
+      case 1 => Left
+      case 2 => Down
+      case 3 => Up
+      case _ => throw new RuntimeException("unknown direction index")
+    }
+  }
+
+  private case object Right extends Direction {
+    override def toInt: Int = 0
+  }
+
+  private case object Left extends Direction {
+    override def toInt: Int = 1
+  }
+
+  private case object Down extends Direction {
+    override def toInt: Int = 2
+  }
+
+  private case object Up extends Direction {
+    override def toInt: Int = 3
+  }
+
+  /** the segmentation label */
+  case class Label(value: Int) extends AnyVal
+
+  /** binary distribution, local information */
+  type BinaryLabelDistribution = PixelImage[(Label, Label) => Double]
+
+  private type MessageFieldBuf = ImageBuffer[Array[LabelDist]]
+
+  /** MRF segmentation with Loopy Belief Propagation
+    * Builds simple Gaussian color models for each label iteratively
+    *
+    * @param image image to segment
+    * @param labelInit initial label where known
+    * @param numLabels number of labels in use */
+
+  def segmentImage(image: PixelImage[RGB],
+                   labelInit: PixelImage[Option[Label]],
+                   binaryDistribution: PixelImage[(Label, Label) => Double],
+                   numLabels: Int, numIterations: Int,
+                   gui: Boolean = false): PixelImage[LabelDist] = {
+    require(labelInit.values.forall(i => i.forall(_.value < numLabels)))
+
+    // init: color models and start labels
+    val initialSegmentation = labelInit.map{init => LabelDist.fromSingleLabel(init.getOrElse(Label(numLabels-1)), numLabels, 0.01)}
+    var colorDists = estimateColorDists(image, initialSegmentation)
```

Comment:
Again naming dist can be ambiguous and stand for both distance and distribution. The code profit from resolving this ambiguity.